src/economy/request.cc

```diff
@@ -273,72 +274,67 @@ Time Request::get_required_time() const {
-			}
-		}
-	}
-
-	if (cost > PRIORITY_MAX_COST) {
-		cost = PRIORITY_MAX_COST;
+uint32_t Request::get_priority(const int32_t cost) const {
+	assert(cost >= 0);
+	const WarePriority& priority =
+	   (target_building_ ? target_building_->get_priority(get_type(), get_index()) :
+	                       WarePriority::kNormal);
+
+	if (WarePriority::kVeryHigh <= priority) {
+		// Always serve requests with the highest priority first,
+		// even if other requests have to wait then.
+		return std::numeric_limits<uint32_t>::max();
+	} else if (priority <= WarePriority::kVeryLow) {
+		// Requests with priority 0 are processed by the
+		// Economy only if there are no other requests.
+		return 0;
 	}
 
-	// priority is higher if building waits for ware a long time
-	// additional factor - cost to deliver, so nearer building
-	// with same priority will get ware first
-	//  make sure that idle request are lower
-	return MAX_IDLE_PRIORITY +
-	       std::max(
-	          uint32_t(1),
-	          ((economy_->owner().egbase().get_gametime().get() -
-	            (is_construction_site ? get_required_time().get() : get_last_request_time().get())) *
-	              WAITTIME_WEIGHT_IN_PRIORITY +
-	           (PRIORITY_MAX_COST - cost) * COST_WEIGHT_IN_PRIORITY) *
-	             modifier);
+	const uint32_t cur_time = economy_->owner().egbase().get_gametime().get();
+	const uint32_t req_time =
+	   (target_constructionsite_ ? get_required_time().get() : get_last_request_time().get());
+	return
+	   // Linear scaling of request priority depending on
+	   // the building's user-specified ware priority.
+	   priority.to_weighting_factor() *
+	   // Linear scaling of request priority depending on the time
+	   // since the request was last supplied (constructionsites)
+	   // or when the next ware is due (productionsites)
+	   (cur_time > req_time ? cur_time - req_time : 1) *
+	   // Requests with higher costs are preferred to keep the average waiting
+	   // times short. This is capped at an arbitrary max cost of 30 seconds
+	   // gametime to not disadvantage close-by supplies too much.
+	   std::max(1, 30000 - cost);
 }
 
 /**
- * Return the transfer priority, based on the priority set at the destination
+ * Return the transfer priority, based on the priority set at the destination,
+ * normalized on a scale from 0 (lowest) to Flag::kMaxTransferPriority (highest).
  */
-// TODO(sirver): Same comment as for Request::get_priority.
-uint32_t Request::get_transfer_priority() const {
-	uint32_t pri = 0;
-
-	if (target_building_) {
-		pri = target_building_->get_priority(get_type(), get_index());
-		if (target_constructionsite_) {
-			return pri + 3;
-		} else if (target_warehouse_) {
-			return pri - 2;
-		}
+uint32_t Request::get_normalized_transfer_priority() const {
+	if (!target_building_) {
+		return 0;
+	}
+
+	// Magic numbers for reasonable weighting. Results for the values at the time of writing:
```

Comment:
This comment does not match the values in widelands.cc - maybe refer to them here rather than stating numbers that might outdate?